954 admiral strategy (#1068)
pharmaverse/admiral: vignettes/programming_strategy.Rmd

# Functions in R

## Function Design

Firstly, it is important to explain how we decide on the need for new derivation functions. 

If a derivation rule or algorithm is common and highly similar across different variables/parameters
(e.g. study day or duration) then we would provide a generic function that can be used to satisfy all
the times this may be needed across different ADaMs. Similarly, if we feel that a certain derivation
could be useful beyond a single purpose we also would provide a generic function (e.g. instead of a
last known alive date function, we have an extreme date function where a user could find the last date
from a selection, or for example the first).

> Otherwise if we feel that a derivation rule is a unique need or sufficiently complex to justify then we

Reviewer: ```suggestion
Otherwise, if we feel that a derivation rule is a unique need or sufficiently complex to justify then we
```

Author: agree

opt for a dedicated function for that specific variable/parameter (e.g. treatment-emergent flag for AEs).

If certain variables are closely connected (e.g. an imputed date and the corresponding imputation flag)
then a single function would provide both variables.

If something needed for ADaM could be achieved simply via an existing tidyverse function, then we do not
wrap this into an admiral function, as that would add an unnecessary extra layer for users.

> The following principles are key when designing a new function:

Reviewer: Should these line up with the the four principles in the README?

* Usability
* Simplicity
* Findability 
* Readability

Reviewer: I guess I see them being the same.  The functions are the package and we have the 4 design principles that influence our practical approach to designing the functions.  Anyways, I am partial to short declarative sentences for designing stuff - as it is something I repeat to myself as I code things... e.g. "one proc away".  If you both are happy with how it reads, then I don't really have an issue.

Reviewer 2: I see. Thanks Ben. We still have the whole "how to program a function, step by step" issue in the backlog. I think when we get to that we might need to adjust some wording as well :) I resolve this for now.

* _**Modularity**_ - All code follows a modular approach, i.e. the steps must be clearly separated and
have a dedicated purpose. This applies to scripts creating a dataset where each module should create a
single variable or parameter. But also to complex derivations with several steps. Commenting on these
steps is key for readability.

* _**Avoid Copy and Paste**_ - If the same or very similar code is used multiple times, it should be put
into a separate function. This improves readability and maintainability and makes unit testing easier.
This should not be done for every simple programming step where tidyverse can be used. But rather for
computational functions or data checks. However, also consider not to nest too many functions.

* _**Checks**_ - Whenever a function fails, a meaningful error message must be provided with a clear

> reference to the input which caused the failure. This should avoid the users have to dig into detailed
> code if they only want to apply a function. This supports usability.

Reviewer: ```suggestion
reference to the input which caused the failure. A users should not have to dig into detailed
code if they only want to apply a function.  A meaningful error message supports usability.
```
